pkg/server/home_workspaces.go

```diff
@@ -0,0 +1,661 @@
+	if !workspaceName.HasPrefix(h.homePrefix) {
+		return
+	}
+	length := strings.Count(workspaceName.String(), ":")
+	homeRootWorkspaceLength := strings.Count(h.homePrefix.String(), ":")
+	lastBucketWorkspaceLength := homeRootWorkspaceLength + h.bucketLevels
+	homeWorkspaceLength := lastBucketWorkspaceLength + 1
+	switch {
+	case length <= homeRootWorkspaceLength:
+		return
+	case length <= lastBucketWorkspaceLength:
+		return true, "Homebucket"
+	case length == homeWorkspaceLength:
+		return true, "Home"
+	}
+	return
+}
+
+// checkWorkspaceLocally checks whether a workspace is known on the current shard
+// in local informers.
+// For home workspaces, we also check:
+//   - if related RBAC resources are there,
+//   - if the workspace phase is READY
+// and if not answer to retry later.
+func (h *homeWorkspaceHandler) checkWorkspaceLocally(workspaceName logicalcluster.Name, isHome bool, userName string) (found, retryLater bool, retryAfterSeconds int, checkError error) {
+	if h.mocks.checkWorkspaceLocally != nil {
+		return h.mocks.checkWorkspaceLocally(workspaceName, isHome, userName)
+	}
+
+	localWorkspace, checkError := h.getLocalClusterWorkspace(workspaceName)
+	if checkError != nil {
+		return found, retryLater, retryAfterSeconds, checkError
+	}
+	if localWorkspace == nil {
+		return found, retryLater, retryAfterSeconds, checkError
+	}
+
+	if workspaceUnschedulable(localWorkspace) {
+		// The workspace cannot be scheduled => return the error.
+		checkError = kerrors.NewForbidden(tenancyv1alpha1.SchemeGroupVersion.WithResource("clusterworkspaces").GroupResource(), localWorkspace.Name, errors.New("unschedulable workspace cannot be accessed"))
+		return found, retryLater, retryAfterSeconds, checkError
+	}
+
+	found = true
+	if !isHome {
+		return found, retryLater, retryAfterSeconds, checkError
+	}
+
+	if localWorkspace.Status.Phase != tenancyv1alpha1.ClusterWorkspacePhaseReady {
+		// We have to wait for the workspace to be Ready before allowing actions in it,
+		// but only for the the home workspaces.
+		// Waiting for home buckets to be ready is done in the `tryToCreate` function.
+
+		// TODO (david): For now we accept all requests in this workspace when it's initializing...
+		// TODO (david): In the future, we might have a way to identify requests that are done through the
+		// initializer virtual workspace
+		// Or it might be unnecessary since other requests will always be blocked until initialization is finished.
+		if localWorkspace.Status.Phase != tenancyv1alpha1.ClusterWorkspacePhaseInitializing {
+			retryLater = true
+			retryAfterSeconds = h.creationDelay
+			return found, retryLater, retryAfterSeconds, checkError
+		}
+	}
+
+	if userName == kuser.APIServerUser {
+		// if request user is system:apiserver, don't wait for the RBAC resources to be setup
+		// (the system:apiserver user request is probably being creating those RBAC resources)
+		return found, retryLater, retryAfterSeconds, checkError
```

Comment:
who is system:apiserver in our case?

Reply:
will go away when making home owner admin

Reply:
Done in commit 0cb62c3aa30411ef67a7350b9ab85dec02c3e210